Update README.md (#15)
patientempowerment/empower-frontend: README.md

# EMPOWER

An app that visualizes the effects of changing inputs on the results of given logistic regression outputs.
It can be used for any kind of clean data. For the example configuration, we use a health use case to predict tuberculosis, diabetes, asthma and COPD on life style choices and patient parameters like age, weight, etc.

## Steps to run
1. Clone the project
2. Configure

>     * Connect to a running server of [patientEmpowerment](https://github.com/KBorchar/patientEmpowerment) via editing `assets/server.conf`, then load `assets/feature-config.json` and `assets/models.json` from there. 

Reviewer: change URL to code in organization repo once migrated

Reviewer: ```suggestion
    * Connect to a running server of [patientEmpowerment](https://github.com/patientempowerment/patientEmpowerment) via editing `assets/server.conf`, then load `assets/feature-config.json` and `assets/models.json` from there. 
```

```
{
  "asthma": {
    "features": {
      "age": {
        "coef": 0.08657563511716311, 
        "mean": 56.90638617580766
      },
      ...
    }
  },
  ...
}
```
First key is the label, in this case "asthma". "features" contains a list of all respective features, their coefficients, and their means.  This file is auto-generated, whenever the backend fits the corresponding model.

#### labels.conf
`assets/labels.conf` specifies all labels to predict for, and their user-facing names. It is used to query the models.json from the server.
With the key "label_titles", you need to set the corresponding title to a label name.

```
{
  "labels": [
  	"asthma",
   ...
   ],
  "label_titles": {
    "asthma": "Asthma",
    ...
  }
}
```

## How to use
There are two visual representations of the same concept: A bar chart (default), and a more interactive prototype. Swipe right on the bar chart screen to access it.
Should any inputs be empty, the app will assume the corresponding feature's mean (from `assets/models.json`).

### Bars
![Alt text](/assets/bars_prototype.png "Bars Prototype")

> On the left side, users can input their data and press the button in the middle to get the output. The output reacts in real time to any changes on the input. You can hide it again by pressing the same button in the middle. To reset all inputs to their means, press the floating button on the bottom right.

Reviewer: ```suggestion
On the left side, users can input their data and press the button in the middle to get the output. The output reacts in real time to any changes on the input. You can hide it again by pressing the same button in the middle. To reset all inputs to their means, press the floating action button on the bottom right.
```

### Bubbles
![Alt text](/assets/bubbles_prototype.png "Bubbles Prototype")

> In the 'Bubbles' prototype the input fields have a bubble representation. They are initially located at the top of the screen. The output is represented by fixed bubbles, arranged around the center image. Each output bubble represents one label to predict on. Users can input their data by dragging or tapping an input bubble, upon which an input dialog opens. By pressing "ok" the change is propagated to the output bubbles and their size changes according to the new probabilities. Additionally, there are particles, small bubbles, flying from the input to the influenced output bubble(s) to indicate their correlation.

Reviewer: ```suggestion
In the 'Bubbles' prototype the input fields have a bubble representation. They are initially located at the top of the screen. The output is represented by fixed label bubbles, arranged around the center image. Each label bubble represents one label to predict on. Users can input their data by dragging or tapping an input bubble, upon which an input dialog opens. By pressing "ok" the change is propagated to the label bubbles and their size changes according to the new probabilities. Additionally, there are particles, small bubbles, flying from the input to the influenced label bubble(s) to indicate their correlation.
```

# EMPOWER

> An app that visualizes the effects of changing inputs on the results of given logistic regression outputs.

Reviewer 2: @justus-hildebrand

Reviewer 3: ```suggestion
Empower is an app that shows the relations between changes in feature vectors and the resulting effects on predictions made by pre-trained logistic regressors. It does this by providing users with an input interface to make changes to feature values and on the fly presenting the resulting effects on the model's prediction results.
```

Author: That is sooo complicated..It might be accurate, bu probably good as second sentence? So like "visualization" first and then "To be more exact <your part>".

> It can be used for any kind of data. For the example configs, we used a health use case to predict tuberculosis, diabetes, asthma and COPD on certain parameters like age, weight, height and smoking behavior.

Reviewer 2: ```suggestion
It can be used for any kind of clean data. For the example configuration, we use a health use case to predict tuberculosis, diabetes, asthma and COPD on life style choices and patient parameters like age, weight, etc.
```

## Steps to run
1. Clone the project
2. Configure

>     * Connect to a running server of [patientEmpowerment](https://github.com/KBorchar/patientEmpowerment) via editing `assets/server.conf` and load `assets/feature-config.json` and `assets/models.json` from there. 

Reviewer 2: Are we migrating these to patientempowerment?

Author: Do you mean migrating the backend repo? I'd vote yes.

Reviewer 2: k then this needs a TODO

>     * Connect to a running server of [patientEmpowerment](https://github.com/KBorchar/patientEmpowerment) via editing `assets/server.conf` and load `assets/feature-config.json` and `assets/models.json` from there. 

Reviewer 2: ```suggestion
    * Connect to a running server of [patientEmpowerment](https://github.com/KBorchar/patientEmpowerment) via editing `assets/server.conf`, then load `assets/feature-config.json` and `assets/models.json` from there. 
```

    * OR:

>     * To configure the input and output manually, edit `assets/feature-config.json` and `assets/models.json` as described further down.

Reviewer 2: ```suggestion
    * To configure the input and output manually, edit `assets/feature-config.json` and `assets/models.json` (see Configuration).
```

> The example files were created with the [backend](https://github.com/KBorchar/patientEmpowerment).

Reviewer 2: ```suggestion
The example files were created by the [backend](https://github.com/KBorchar/patientEmpowerment).
```

> 3. To configure the displayed titles of the output, edit `assets/labels.conf`.

Reviewer 2: ```suggestion
3. To configure the user-facing names of the output labels, edit `assets/labels.conf`.
```

4. Run the app in an IDE of your choice (e.g. [Android Studio with Flutter Plugin](https://androiddvlpr.com/flutter-android-studio/)).

### Configuration
#### feature-config.json
The content of the feature config should look similar to this:
```
{
  "copd": {
    "title": "COPD",
    "choices": {
      "Yes": 1,
      "No": 0
    }
  },
  "age": {
    "title": "Age",
    "slider_min": 40,
    "slider_max": 70
  },
  ...
}
```
It represents the possible input fields.

> As key, you use the feature name, give it a title and decide whether it should be represented as a slider or radio choice buttons. For radio buttons, specify a list of available choices with their correlated number representation in your ML algorithm. For a slider, specify the minimal and maximal value that the user should be able to input.

Reviewer 2: ```suggestion
As key, you use the feature name. Give it a user-facing title and decide whether it should be represented as a slider or choice. For choices, specify a list of available choices with their correlating value in the original data set. For a slider, specify the minimal and maximal value that the user should be able to input.
```

#### models.json

> In `models.json` you specify, per label, the ML determined coefficients for each feature as well as their means. It should look similar to this:

Reviewer 2: ```suggestion
In `models.json` you specify, per label, the coefficients for each feature as well as their means in the original data set. It should look similar to this:
```

```
{
  "asthma": {
    "features": {
      "age": {
        "coef": 0.08657563511716311, 
        "mean": 56.90638617580766
      },
      ...
    }
  },
  ...
}
```

> First key is the label name, in this case "COPD". inside the "feature" argument you then can specify a list of features that influence this label with the respective coefficients (at key "coeff") and means (at key "mean"). Repeat this for all features and labels.

Reviewer 2: ```suggestion
First key is the label, in this case "asthma". "features" contains a list of all respective features, their coefficients, and their means.
```

Reviewer 2: ```suggestion
First key is the label name, in this case "COPD". inside the "feature" argument you then can specify a list of features that influence this label with the respective coefficients (at key "coeff") and means (at key "mean"). Repeat this for all features and labels. This file is auto-generated, whenever the [backend] fits the corresponding model.
```

#### labels.conf

> In `assets/labels.conf` you specify all labels that your want to predict for, that also have a representation in `assets/models.json` (!) as a simple list with the key "labels".

Reviewer 2: ```suggestion
`assets/labels.conf` specifies all labels to predict for, and their user-facing names.
```

```
{
  "labels": [
  	"asthma",
   ...
   ]
}
```

## How to use

> There are two visual representations of the same concept: Bars on startup and Bubbles once you swipe one page to the right.

Reviewer 2: ```suggestion
There are two visual representations of the same concept: A bar chart (default), and a more interactive prototype. Swipe right on the bar chart screen to access it.
```

> Should you have left out some input parameters, the app will assume the means given in `assets/models.json`.

Reviewer 2: ```suggestion
Should any inputs be empty, the app will assume the corresponding feature's mean (from `assets/models.json`).
```

### Bars
![Alt text](/assets/bars_prototype.png "Bars Prototype")

> On the left side, you can put in your data and press the button in the middle to get the output. You can also change the input while the output is displayed or hide the output bars again by pressing the same button in the middle. To reset to means, press the floating button on the bottom right.

Reviewer 2: ```suggestion
On the left side, users can input their data and press the button in the middle to get the output. The output reacts in real time to any changes on the input. You can hide it again by pressing the same button in the middle. To reset all inputs to their means, press the floating button on the bottom right.
```

### Bubbles
![Alt text](/assets/bubbles_prototype.png "Bubbles Prototype")

> In Bubbles Prototype the input fields have a bubble representation. They are located at the top of the screen. The output is also in a circular form in the screen center around a center image. Each output bubble representing one ML prediction label. You can input your data by dragging or clicking an input bubble. A dialog opens and asks you to input your data. By pressing "ok" the change is propagated to the output bubbles and their size changes according to the new probabilities. Additionally there are particles, small bubbles, flying from the input to the influenced output bubble(s) to indicate their correlation.

Reviewer 2: ```suggestion
In the 'Bubbles' prototype the input fields have a bubble representation. They are initially located at the top of the screen. The output is represented by fixed bubble, arranged around the center image. Each output bubble represents one label to predict on. Users can input their data by dragging or tapping an input bubble, upon which an input dialog opens. By pressing "ok" the change is propagated to the output bubbles and their size change according to the new probabilities. Additionally, there are particles, small bubbles, flying from the input to the influenced output bubble(s) to indicate their correlation.
```

Author: Just fixed some typos
```suggestion
In the 'Bubbles' prototype the input fields have a bubble representation. They are initially located at the top of the screen. The output is represented by fixed bubbles, arranged around the center image. Each output bubble represents one label to predict on. Users can input their data by dragging or tapping an input bubble, upon which an input dialog opens. By pressing "ok" the change is propagated to the output bubbles and their size changes according to the new probabilities. Additionally, there are particles, small bubbles, flying from the input to the influenced output bubble(s) to indicate their correlation.
```

## Contributing
We are happy if you'd want to contribute to the project. To get into the development, you might want to take some things into account.

> ### Naming

Reviewer 2: honestly I would remove the whole # Naming section

Author: I just thought it might be confusin to people not familiar with ML language.

Reviewer 2: Then they shouldn't use this app :) <3

Author: Sure. :D
